### PR TITLE
[DTensor] min/max dim reduction fix 2

### LINKED ISSUE #168940: [DTensor] aten.max.dim returns wrong indices when using DTensor
### 🐛 Describe the bug

I found that current strategy of `aten.max.dim` may get incorrect indices output if sharded the dim for maximization.

Sample code:
```python
import torch
from torch.distributed.tensor import distribute_tensor, Shard

from torch.testing._internal.common_utils import run_tests
from torch.testing._internal.distributed._tensor.common_dtensor import DTensorTestBase, with_comms


class TestRegisterSharding(DTensorTestBase):
    @with_comms
    def test_max_dim(self):
        mesh = self.build_device_mesh()

        x = torch.randn(4, 4, device="cuda")

        max_value, max_indices = torch.max(x, dim=1)

        dist_x = distribute_tensor(x, mesh, [Shard(1)])

        dist_max_value, dist_max_indices = torch.max(dist_x, dim=1)

        print("x:", x)
        print("max_value:", max_value)
        print("max_indices:", max_indices)
        print("dist_max_value:", dist_max_value.full_tensor())
        print("dist_max_indices:", dist_max_indices.full_tensor())


if __name__ == "__main__":
    run_tests()
```

Result:
```python
x: tensor([[-1.6165,  0.5685, -0.5102, -0.9113],
        [-1.1555, -0.2262, -1.2891,  1.0654],
        [-0.7167, -0.5333,  0.2078, -0.9798],
        [ 0.7447, -0.2395,  0.2737,  0.0920]], device='cuda:0')
max_value: tensor([0.5685, 1.0654, 0.2078, 0.7447], device='cuda:0')
max_indices: tensor([1, 3, 2, 0], device='cuda:0')
dist_max_value: tensor([0.5685, 1.0654, 0.2078, 0.7447], device='cuda:0')
dist_max_indices: tensor([0, 0, 0, 0], device='cuda:0')
```

Each rank gets a shape(4, 1) local tensor to call `max.dim` in this case, and the local result of max indices is [0, 0, 0, 0]. The framework doesn't process the offset of index, which leads to an incorrect global result when the relavant dim is sharded.

Is there a good way to implement a strategy that supports sharding the index dim?

### Versions

torch v2.9.0

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @tianyu-l @XilunWu @SherlockNoMad

### PR DESCRIPTION
**Pull request by skpark-rh**
Fixes #168940

## Issue
The bug happens because min/max with dim reduction is nonlinear.  The returned index tensor is the culprit.

## solution
Created custom handler to compute local reductions with global indices.  Use the min/max values to perform a linear reduction locally.

### CODE REVIEW

**Comment by skpark-rh** [comment]
Hey @zpcore sorry about this... Something went horribly wrong when I merged main into this branch. I resolved most of the comments. I'll repost the answer to your question here.

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+        placements: tuple[Placement, ...],
+    ) -> tuple[int, torch.Tensor]:
+        local_shape, global_offset = compute_local_shape_and_global_offset(
+            global_shape, device_mesh, placements
+        )
+
+        if self._dim is None:
+            local_coord = torch.unravel_index(local_idx, local_shape)
+            global_coord = torch.stack(local_coord)
+            gather_dim = 0
```
Comment (skpark-rh) [inline torch/distributed/tensor/_nonlinear_redux.py]: > Why do we return `gather_dim=0` here?

For `dim = None`, the tensor is flattened and a single value/index is returned. I put the `gather_dim=0` to collect all the operated values from different processes to the 0th dimension. I call the `op_call` again in `__call__` of the child classes to get the rank winners on the 0th dim to get the true winning number.

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+            for i, offset in enumerate(global_offset):
+                global_coord[i] += offset
+            # compute with proper striding
+            gathered_idxs = torch.tensor(0, device=local_idx.device, dtype=torch.long)
+            for i, coord in enumerate(global_coord):
+                gathered_idxs += coord * reduce(operator.mul, global_shape[i + 1 :], 1)
```
Comment (skpark-rh) [inline torch/distributed/tensor/_nonlinear_redux.py]: > This should also work. I think your code is also fine though since `global_shape` is not large.
>```python
>accum_size = 1
>for coord, dim_size in reversed(zip(global_coord, global_shape)):
>    gathered_idxs += coord * accum_size
>    accum_size *= dim_size
>```

Yes, since I knew that `global_shape` cannot be large, I went with this reduce operation.

File: torch/distributed/tensor/_nonlinear_redux.py (hunk 1 above)
Comment (zpcore) [reply torch/distributed/tensor/_nonlinear_redux.py]: I see! Now I know how you gathered those min/max value below by assuming its dim 0.

**Review by zpcore — APPROVED** [review]
LGTM! Thanks for improving the op coverage

**Comment by ezyang** [comment]
The reenabled test here is flaking on master so I am reverting this.

File: torch/distributed/tensor/_dispatch.py [hunk 3]
```diff
@@ -154,8 +154,10 @@ def __init__(self) -> None:
             aten.convolution_backward.default: convolution_backward_handler,
             aten._amp_foreach_non_finite_check_and_unscale_.default: found_inf_reduce_handler,
             aten.as_strided.default: as_strided_handler,
-            aten.argmin.default: argmin_argmax_handler,
-            aten.argmax.default: argmin_argmax_handler,
+            aten.argmin.default: ArgMinMaxHandler(),
+            aten.argmax.default: ArgMinMaxHandler(),
+            aten.max.dim: MinMaxDimHandler(),
+            aten.min.dim: MinMaxDimHandler(),
```
Comment (skpark-rh) [inline torch/distributed/tensor/_dispatch.py]: @ezyang I believe this is where the process group race condition is happening... Putting a factory function to create a new handler classes at every call to the min/max/argmin/argmax class will isolate the object from the different PGs preventing race condition. What do you think?

Comment (ezyang) [reply torch/distributed/tensor/_dispatch.py]: I skimmed the PR and I don't understand why we are doing OO here at all. What's the reason?

Comment (skpark-rh) [reply torch/distributed/tensor/_dispatch.py]: I wanted to use inheritance to make the code reusable. The min/max and argmin/argmax require similar `__calls__` but still different enough to warrant two separate semantics.  Doing it through OO helps layout and organization.

File: torch/distributed/tensor/_nonlinear_redux.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    def impl(
+        op_call: torch._ops.OpOverload,
+        args: tuple["dtensor.DTensor", int] | tuple["dtensor.DTensor", int, bool],
+        kwargs: dict[str, object],
+    ):
+        # create a new handler
+        handler = MinMaxDimHandler()
+        return handler(op_call, args, kwargs)
+
+    return impl
```
Comment (skpark-rh) [inline torch/distributed/tensor/_nonlinear_redux.py]: @ezyang @zpcore Added factory method to prevent race conditions.  Ran same tests multiple times to verify.

File: torch/distributed/tensor/_dispatch.py (hunk 3 above)
Comment (ezyang) [reply torch/distributed/tensor/_dispatch.py]: If you don't actually need an object with instance data, you can still do inheritance with `@classmethod`. Why do you use self?

Comment (skpark-rh) [reply torch/distributed/tensor/_dispatch.py]: I need an object with instance data here. I am keeping track of `dim`, `keepdim`, `shard_mesh_list`, and `op_call_repr`.  I also am careful using classmethods as they can muck up the inheritance/overrides.   You are absolutely right that I can make this code not rely on instance data and purely keep it stateless as a factory method with classmethods but I thought using self was simpler.

**Comment by skpark-rh** [comment]
I am curious how it was able to push to my fork though...

**Review by zpcore — APPROVED** [review]
LGTM! Thanks
